File: src/test/java/siosio/it/MemoResourceTest.java
package siosio.it;

import static org.junit.Assert.assertThat;

import java.io.File;
import java.util.List;

import org.hamcrest.collection.IsEmptyCollection;
import org.jboss.arquillian.container.test.api.Deployment;
import org.jboss.arquillian.extension.rest.client.ArquillianResteasyResource;
import org.jboss.arquillian.junit.Arquillian;
import org.jboss.shrinkwrap.api.ShrinkWrap;
import org.jboss.shrinkwrap.api.asset.EmptyAsset;
import org.jboss.shrinkwrap.api.spec.WebArchive;
import siosio.memo.MemoForm;
import siosio.memo.MemoResource;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(Arquillian.class)
public class MemoResourceTest {

    @Deployment
    public static WebArchive createDeployment() {
        return ShrinkWrap.create(WebArchive.class, "memomemo.war")
                .addAsResource("META-INF/persistence.xml")
                .addPackages(true, "siosio");
    }

    @Test
    public void test(@ArquillianResteasyResource MemoResource resource) throws Exception {
        final List<MemoForm> result = resource.list();

        assertThat(result, IsEmptyCollection.<MemoForm>empty());
    }
}

File: src/test/java/siosio/it/MemoResourceTest.kt
package siosio.it

import org.eclipse.persistence.jaxb.rs.*
import org.glassfish.hk2.api.*
import org.glassfish.hk2.utilities.binding.*
import org.glassfish.jersey.client.*
import org.glassfish.jersey.server.*
import org.glassfish.jersey.test.*
import org.hamcrest.collection.*
import org.hamcrest.core.*
import org.junit.*
import siosio.entity.*
import siosio.memo.*
import siosio.service.*
import java.util
import java.util.*
import javax.ws.rs.core.*
import kotlin.properties.*

public class MemoResourceTest : JerseyTest() {

  override fun configure(): Application? {
    enable(TestProperties.DUMP_ENTITY)
    enable(TestProperties.LOG_TRAFFIC)
    return ResourceConfig(MemoResource::class.java)
        .register(object : AbstractBinder () {
          override fun configure() {
            bindFactory(object : Factory<MemoService> {
              override fun dispose(instance: MemoService?) {
              }

              override fun provide(): MemoService? {
                return MemoServiceMock()
              }
            }).to(MemoService::class.java)
          }
        })
  }

  override fun configureClient(config: ClientConfig) {
    config.register(MOXyJsonProvider::class.java)
  }

  class MemoServiceMock : MemoService() {
    companion object {
      var findAllResult: List<MemoEntity> by Delegates.notNull()
    }
    override fun findAll(): List<MemoEntity> {
      return findAllResult
    }
  }

  @Test
  fun `データが存在しない場合からのリストが戻されること`() {
    MemoServiceMock.findAllResult = listOf()
    val result = target("memo")
        .request()
        .accept(MediaType.APPLICATION_JSON_TYPE)
        .get()
    Assert.assertThat(result.status, Is.`is`(200))
    Assert.assertThat(result.mediaType, Is.`is`(MediaType.APPLICATION_JSON_TYPE))
    val entity = result.readEntity(object : GenericType<ArrayList<MemoForm>>() {})
    Assert.assertThat(entity, IsEmptyCollection.empty())
  }
}